cpp/src/tracing.cpp

```diff
@@ -0,0 +1,240 @@
+};
+
+void initTracer(std::unique_ptr<sdktrace::SpanExporter> exporter)
+{
+  // auto exporter = std::unique_ptr<sdktrace::SpanExporter>(
+  //     new opentelemetry::exporter::trace::OStreamSpanExporter);
+  auto processor = std::unique_ptr<sdktrace::SpanProcessor>(
+      new sdktrace::SimpleSpanProcessor(std::move(exporter)));
+  auto provider = nostd::shared_ptr<opentelemetry::trace::TracerProvider>(
+      new sdktrace::TracerProvider(std::move(processor)));
+
+  // Set the global trace provider
+  opentelemetry::trace::Provider::SetTracerProvider(provider);
+
+  // set global propagator
+  opentelemetry::context::propagation::GlobalTextMapPropagator::
+      SetGlobalPropagator(
+          nostd::shared_ptr<
+              opentelemetry::context::propagation::TextMapPropagator>(
+              new opentelemetry::trace::propagation::HttpTraceContext()));
+}
+
+nostd::shared_ptr<trace::Tracer> get_tracer()
+{
+  auto provider = trace::Provider::GetTracerProvider();
+  nostd::shared_ptr<opentelemetry::v0::trace::Tracer> tracer =
+      provider->GetTracer("qpid-tracer");
+  return tracer;
+}
+
+void on_message_span(message &message, delivery &d) {
+
+    opentelemetry::trace::StartSpanOptions options;
+    options.kind          = opentelemetry::trace::SpanKind::kConsumer;
+
+    // extract context from AMQP message annotations
+    const AMQPTextMapCarrier<proton::message::annotation_map> carrier(
+        &message.message_annotations());
+
+    nostd::shared_ptr<
+        opentelemetry::v0::context::propagation::TextMapPropagator>
+        prop = opentelemetry::context::propagation::GlobalTextMapPropagator::
+            GetGlobalPropagator();
+    context::Context current_ctx =
+        opentelemetry::context::RuntimeContext::GetCurrent();
+
+    context::Context new_context = prop->Extract(carrier, current_ctx);
+    options.parent =
+        opentelemetry::trace::propagation::GetSpan(new_context)->GetContext();
+
+    binary tag_in_binary = d.tag();
+    std::string tag_in_string = std::string(d.tag());
+    std::stringstream ss;
+    for(int i=0; i<(int)tag_in_string.length(); ++i)
+        ss << std::hex << (int)tag_in_binary[i];
+    std::string delivery_tag = ss.str();
+
+    receiver r = d.receiver();
+    source s = r.source();
+    std::string s_addr = s.address();
+
+    transfer tt(d);
+    std::string delivery_state = to_string(tt.state());
+
+    nostd::shared_ptr<trace::Span> span =
+        get_tracer()->StartSpan("amqp-message-received",
+                                {{"Delivery_tag", delivery_tag},
+                                 {"Source_address", s_addr}},
+                                options);
+
+    trace::Scope scope = get_tracer()->WithActiveSpan(span);
+
+    if (delivery_state == "unknown")
+    {
+        span->AddEvent("Delivery state is unknown");
+    }
+    if (delivery_state == "Received")
+    {
+        span->AddEvent("Delivery is received");
+    }
```

Comment:
A delivery state is unlikely to ever be known at this point (creating the span, before firing on_message), the client or application will typically set it once they are given the message and finish processing it. That may be before on_message returns if they do it explicitly, or they may have the 'auto-accept' enabled (it is by default) and it is done by the client just after on_message returns, or they may disable that and do it manually later after some long processing completes.

For the JMS client the same selection of possibilities (and others) are true, and we cant easily access the state or span later anyway (and cant leave it open forever either, and need to create a new active span for the next message delivery). We also have cases there that the proton clients do not, where we specifically *dont* deliver the message to the application.

So what I did was, when onMessage returns (or the message wasnt given to onMessage at all) and I am completing the delivery span, I simply logged a summary event relating to the onMessage invocation, not the literal delivery state as it may well not be known yet, adding a description of what happened with e.g 'event = DELIVERED' (or EXPIRED, REDELIVERIES_EXCEEDED...or APPLICATION_ERROR if they throw an exception from onMessage).